Bugfixes and other changes

Xento opened the pull request.

drndos commented on app.py:
```diff
@@ -78,12 +79,20 @@ def spool_info():
+    
+    issue = False
+    #TODO: Fix issue when external spool info is reset via bambulab interface
+    augmentTrayDataWithSpoolMan(spool_list, vt_tray_data, trayUid(EXTERNAL_SPOOL_AMS_ID, 254))
```
Please use constant instead of hard coded 254

Xento replied:
I will do a test print to view the original messages from the printer

drndos commented on app.py:
```diff
@@ -161,10 +174,10 @@ def home():
     issue = False
     #TODO: Fix issue when external spool info is reset via bambulab interface
-    augmentTrayDataWithSpoolMan(spool_list, vt_tray_data, trayUid(EXTERNAL_SPOOL_AMS_ID, vt_tray_data["id"]))
+    augmentTrayDataWithSpoolMan(spool_list, vt_tray_data, trayUid(EXTERNAL_SPOOL_AMS_ID, 254))
```
same here

drndos commented on mqtt_bambulab.py:
```diff
@@ -44,6 +44,8 @@ def on_message(client, userdata, msg):
         if ams_used:
           spendFilaments(zip(ams_mapping, expected_filaments_usage))
+        else:
+          spendFilaments(zip([254], expected_filaments_usage))
```
same here

drndos commented on spoolman_service.py:
```diff
@@ -21,13 +30,14 @@ def augmentTrayDataWithSpoolMan(spool_list, tray_data, tray_id):
       #TODO: hardcoded ams_id
       if ams_usage.get(trayUid(0, tray_id)):
         ams_usage[trayUid(0, tray_id)] += float(usage)
+      elif tray_id == 254:
```
also here and below

drndos commented on templates/base.html:
```diff
@@ -37,7 +38,25 @@
 </header>
+<script>
+document.addEventListener("DOMContentLoaded", function () {
+  function attachLinkListeners() {
```
Could you please explain this method in a comment, so I understand what is its purpose ?

Xento replied:
This loads the site with an request and replaces the current site. So in webapp mode it won't show the navigation. 
Maybe this doesn't work when using more JavaScript on the page. 
Maybe there is an better method to build a webapp. 
But this works so far.

drndos commented on templates/fragments/tray.html:
```diff
@@ -5,9 +5,13 @@
+	Empty
     {% endif %}
-    Tray {{ tray_data.id }}
+	{% if (tray_data.id)|int != 254 %}
```
Please use constant

drndos reviewed (commented):
Thank you for the code

weakfl commented:
@Xento tried your PR but the time conversion sometimes fails with:

```
Traceback (most recent call last):
  File "/home/app/app.py", line 201, in home
    augmentTrayDataWithSpoolMan(spool_list, tray, trayUid(ams["id"], tray["id"]))
  File "/home/app/spoolman_service.py", line 25, in augmentTrayDataWithSpoolMan
    dt = datetime.strptime(spool["last_used"], "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=ZoneInfo("UTC"))
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/_strptime.py", line 568, in _strptime_datetime
    tt, fraction, gmtoff_fraction = _strptime(data_string, format)
                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/_strptime.py", line 349, in _strptime
    raise ValueError("time data %r does not match format %r" %
ValueError: time data '2025-03-10T09:19:30.307000Z' does not match format '%Y-%m-%dT%H:%M:%SZ'
```

Seems like the timestamp **can** contain fractional seconds.

Looking at https://github.com/Xento/openspoolman/commit/887da125b5de77a61429d60818fa207d0f8441fc this might have been a problem all along.

Quick (and dirty) fix:

```python
      if "last_used" in spool:
        try:
            dt = datetime.strptime(spool["last_used"], "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=ZoneInfo("UTC"))
        except ValueError:
            dt = datetime.strptime(spool["last_used"], "%Y-%m-%dT%H:%M:%S.%fZ").replace(tzinfo=ZoneInfo("UTC"))
        tz_name = os.getenv("TZ", "Europe/Berlin")
        local_timezone = ZoneInfo(tz_name)
        local_time = dt.astimezone()
        tray_data["last_used"] = local_time.strftime("%d.%m.%Y %H:%M:%S")
```

Xento commented:
> @Xento tried your PR but the time conversion sometimes fails with:
> 
> ```
> Traceback (most recent call last):
>   File "/home/app/app.py", line 201, in home
>     augmentTrayDataWithSpoolMan(spool_list, tray, trayUid(ams["id"], tray["id"]))
>   File "/home/app/spoolman_service.py", line 25, in augmentTrayDataWithSpoolMan
>     dt = datetime.strptime(spool["last_used"], "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=ZoneInfo("UTC"))
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/_strptime.py", line 568, in _strptime_datetime
>     tt, fraction, gmtoff_fraction = _strptime(data_string, format)
>                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/usr/local/lib/python3.11/_strptime.py", line 349, in _strptime
>     raise ValueError("time data %r does not match format %r" %
> ValueError: time data '2025-03-10T09:19:30.307000Z' does not match format '%Y-%m-%dT%H:%M:%SZ'
> ```
> 
> Seems like the timestamp **can** contain fractional seconds.
> 
> Looking at [Xento@887da12](https://github.com/Xento/openspoolman/commit/887da125b5de77a61429d60818fa207d0f8441fc) this might have been a problem all along.
> 
> Quick (and dirty) fix:
> 
> ```python
>       if "last_used" in spool:
>         try:
>             dt = datetime.strptime(spool["last_used"], "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=ZoneInfo("UTC"))
>         except ValueError:
>             dt = datetime.strptime(spool["last_used"], "%Y-%m-%dT%H:%M:%S.%fZ").replace(tzinfo=ZoneInfo("UTC"))
>         tz_name = os.getenv("TZ", "Europe/Berlin")
>         local_timezone = ZoneInfo(tz_name)
>         local_time = dt.astimezone()
>         tray_data["last_used"] = local_time.strftime("%d.%m.%Y %H:%M:%S")
> ```

I never had this issue, but I will try to fix it.

Xento commented:
> @Xento some further observations:
> 
> 1. None of the previously assigned spools seem to be recognized any more. Both A0, and A2 have `active_tray` and `tag` set, but the logs indicates they are not found:
> 
> ```
> AMS [A] (hum: 5, temp: 28.7ºC)
>     - [A0]  3D3E3DFF (-01%) [[ 00000000000000000000000000000000 ]]
>       - Not found. Update spool tag!
>     - [A1] PETG HF ADB1B2FF (095%) [[ 533A92B319DF4C73902F864592F65DE6 ]]
> Skipping set active tray
>     - [A2]  101820FF (-01%) [[ 00000000000000000000000000000000 ]]
>       - Not found. Update spool tag!
>     - [A3] PLA Basic 00AE42FF (084%) [[ 7B653B1C6E5440F7BCF2A66E0B07F5BE ]]
> Skipping set active tray
> ```

I think this message is normal, as this tries to read out the bambulab tags and it seems that you have non bambulab spools in A0 and A2, so it can't find a spool with the tag "00000000000000000000000000000000" in spoolman.

> 2. Clicking on 'Fill' results in an internal server error:
> 
> ```
> Traceback (most recent call last):
>   File "/home/app/venv/lib/python3.11/site-packages/flask/app.py", line 1511, in wsgi_app
>     response = self.full_dispatch_request()
>                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/flask/app.py", line 919, in full_dispatch_request
>     rv = self.handle_user_exception(e)
>          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/flask/app.py", line 917, in full_dispatch_request
>     rv = self.dispatch_request()
>          ^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/flask/app.py", line 902, in dispatch_request
>     return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/app.py", line 76, in fill
>     return render_template('fill.html', spools=spools, ams_id=ams_id, tray_id=tray_id)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/flask/templating.py", line 150, in render_template
>     return _render(app, template, context)
>            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/flask/templating.py", line 131, in _render
>     rv = template.render(context)
>          ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/jinja2/environment.py", line 1295, in render
>     self.environment.handle_exception()
>   File "/home/app/venv/lib/python3.11/site-packages/jinja2/environment.py", line 942, in handle_exception
>     raise rewrite_traceback_stack(source=source)
>   File "/home/app/templates/fill.html", line 1, in top-level template code
>     {% extends 'base.html' %}
>   File "/home/app/templates/base.html", line 67, in top-level template code
>     {% block content %}{% endblock %}
>     ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/templates/fill.html", line 6, in block 'content'
>     {% with action_fill=True %}{% include 'fragments/list_spools.html' %}{% endwith %}
>     ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/templates/fragments/list_spools.html", line 42, in top-level template code
>     <a href="{{ url_for('spool_info', tag_id=spool.extra.get("tag").strip('"')) }}">
>     ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/app/venv/lib/python3.11/site-packages/jinja2/utils.py", line 92, in from_obj
>     if hasattr(obj, "jinja_pass_arg"):
>        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> jinja2.exceptions.UndefinedError: 'None' has no attribute 'strip'
> ```

Is it possible, that this spool does not have a tag assigned?
I will try it by myself later with a spool with no tag.

weakfl commented:
> Is it possible, that this spool does not have a tag assigned?

Yes, some of my spools don't have a tag assigned (yet) and since the `tag` field does not have a default value this seems to result in the property not being set by spoolman, eg:

```json
   {
      "id":15,
      "registered":"2025-02-25T09:59:56Z",
      "filament":{
         "id":15,
         "registered":"2025-02-25T09:58:37Z",
         "name":"Galaxy Purple",
         "vendor":{
            "id":1,
            "registered":"2025-01-24T14:16:10Z",
            "name":"Prusament",
            "external_id":"Prusament",
            "extra":{
               
            }
         },
         "material":"PLA",
         "density":1.24,
         "diameter":1.75,
         "weight":1000.0,
         "spool_weight":193.0,
         "settings_extruder_temp":215,
         "color_hex":"453A72",
         "extra":{
            "type":"\" Basic\"",
            "nozzle_temperature":"[190,300]"
         }
      },
      "remaining_weight":716.0,
      "initial_weight":1000.0,
      "spool_weight":193.0,
      "used_weight":284.0,
      "remaining_length":240063.07150240327,
      "used_length":95220.54791436109,
      "location":"Dry Box",
      "archived":false,
      "extra":{
         "active_tray":"\"\""
      }
   }
```

Xento commented:
Maybe you could try my other branch. https://github.com/Xento/openspoolman/commits/spend-filament-for-local-prints/

There I implemented:
- increased python version to 3.12.9
- logging of the mqtt messages to be able to replay the messages from the printer for better debugging
- tracking of filament usage when printed localy over the display
- fixed fractional seconds in spoolman
- spool info can now be opened by spool id, so no tag has to be assigned

weakfl commented:
@Xento tried your other branch, but now only the external spool is displayed in 'Home', all other spools are missing.

The following error is constantly repeating in the log:

```
Traceback (most recent call last):
  File "/home/app/mqtt_bambulab.py", line 159, in on_message
    if "print" in data:
                  ^^^^
UnboundLocalError: cannot access local variable 'data' where it is not associated with a value
```

If I try to open 'Assign NFC Tag' -> 'Spool info' I get the following error:

```
Traceback (most recent call last):
  File "/home/app/app.py", line 123, in spool_info
    return render_template('spool_info.html', tag_id=tag_id, current_spool=current_spool, ams_data=ams_data, vt_tray_data=vt_tray_data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app/venv/lib/python3.12/site-packages/flask/templating.py", line 150, in render_template
    return _render(app, template, context)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app/venv/lib/python3.12/site-packages/flask/templating.py", line 131, in _render
    rv = template.render(context)
         ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app/venv/lib/python3.12/site-packages/jinja2/environment.py", line 1295, in render
    self.environment.handle_exception()
  File "/home/app/venv/lib/python3.12/site-packages/jinja2/environment.py", line 942, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "/home/app/templates/spool_info.html", line 1, in top-level template code
    {% extends 'base.html' %}
  File "/home/app/templates/base.html", line 76, in top-level template code
    {% block content %}{% endblock %}
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app/templates/spool_info.html", line 5, in block 'content'
    {% include 'fragments/spool_details.html' %}
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app/templates/fragments/spool_details.html", line 10, in top-level template code
    style="background-color: #{{ current_spool.filament.color_hex }}; width: 40px; height: 40px; border-radius: 5px;">
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/app/venv/lib/python3.12/site-packages/jinja2/environment.py", line 490, in getattr
    return getattr(obj, attribute)
           ^^^^^^^^^^^^^^^^^^^^^^^
jinja2.exceptions.UndefinedError: 'None' has no attribute 'filament'
```

Xento commented:
I'm sorry I missed that the conversion from the mqtt message to the data variable got lost.
I think it is fixed now.
Currently i'm implementing a print history with an image and the filament usage.

weakfl commented:
> I'm sorry I missed that the conversion from the mqtt message to the data variable got lost.
I think it is fixed now.

Np and thanks @Xento, first error is indeed fixed and all trays are showing up as expected now.

> If I try to open 'Assign NFC Tag' -> 'Spool info' I get the following error:

This still happen with the trace posted above https://github.com/drndos/openspoolman/pull/20#issuecomment-2723795985

Xento commented:
Should be fixed with https://github.com/Xento/openspoolman/commit/0af9c3fdf955d6890ea7632b806ef98cbfd35f9c
Only a small comparsion error between datatypes.

weakfl commented:
> Should be fixed with [Xento@0af9c3f](https://github.com/Xento/openspoolman/commit/0af9c3fdf955d6890ea7632b806ef98cbfd35f9c) Only a small comparsion error between datatypes.

Can confirm its fixed, thanks.

Xento commented:
Have you tried tracking of local prints?

Maybe you want to try my new branch: https://github.com/Xento/openspoolman/tree/implement-print-history
![image](https://github.com/user-attachments/assets/62df3a5b-9bcd-4f40-9164-d2fe8c770e52)

weakfl commented:
@drndos @Xento `Dockerfile` is broken, needs recursive `mkdir` here:

```diff
-RUN mkdir /home/app/static/prints && chown -R nonroot:nonroot /home/app/static/prints
+RUN mkdir -p /home/app/static/prints && chown -R nonroot:nonroot /home/app/static/prints
```

Thanks for merging and the various improvements though.
